### PR TITLE
fix(nuxt): wrap `new URL()` in try-catch to prevent crashes on invalid input

### PR DESCRIPTION
**Pull request by faisalaffan**
### What

6 locations where `new URL()` could throw unhandled `TypeError` on malformed user input, wrapped with `try/catch`.

---

### 1. `navigateTo()` — open handler (`router.ts:159`)

**Problem**

`toPath` comes directly from user input. A malformed value like `"http://a b.com"` throws raw `TypeError: Invalid URL` with no recovery and no clear message.

**Before**

```ts
const { protocol } = new URL(toPath, window.location.href)
```

After

```ts
let url: URL
try {
  url = new URL(toPath, window.location.href)
} catch {
  throw new Error(`Cannot navigate to invalid URL: '${toPath}'`)
}
const { protocol } = url
```

Solution

Catch the `TypeError` and re-throw as a descriptive `Error` that includes the bad input so the user knows exactly what went wrong.

---

### 2. `navigateTo()` — external URL validation (`router.ts:183`)

**Problem**

Same crash as #1, but on the external-URL code path. No clear message when `toPath` is invalid.

**Before**

```ts
const { protocol } = new URL(toPath, import.meta.client ? window.location.href : 'http://localhost')
```

After

```ts
let url: URL
try {
  url = new URL(toPath, import.meta.client ? window.location.href : 'http://localhost')
} catch {
  throw new Error(`Cannot navigate to invalid URL: '${toPath}'`)
}
const { protocol } = url
```

Solution

Identical pattern to #1. Same `let` + `try/catch` + descriptive `Error`.

---

### 3. `encodeURL()` — internal redirect encoder (`router.ts:344`)

**Problem**

`location` can come from a redirect target string. If it's malformed, the entire redirect logic crashes.

**Before**

```ts
const url = new URL(location, 'http://localhost')
```

After

```ts
let url: URL
try {
  url = new URL(location, 'http://localhost')
} catch {
  return location
}
```

Solution

Return the raw input as-is instead of crashing. The redirect still works, just un-encoded.

---

### 4. `<NuxtLink>` — prefetch path normalization (`nuxt-link.ts:422`)

**Problem**

`path` comes from the link's `to` value. An invalid path crashes prefetch silently in the background, potentially breaking the link render.

**Before**

```ts
const normalizedPath = isExternal.value
  ? new URL(path, window.location.href).href
  : path
```

After

```ts
const normalizedPath = (() => {
  try {
    return isExternal.value
      ? new URL(path, window.location.href).href
      : path
  } catch {
    return path
  }
})()
```

Solution

Wrap in an IIFE so `try/catch` works inside an expression. Fall back to the raw path — prefetch may be suboptimal but nothing crashes.

---

### 5. `<NuxtLink>` — slot route getter (`nuxt-link.ts:530`)

**Problem**

Called during template render. If `href.value` is malformed, the `new URL()` call crashes the entire component, producing a blank page.

**Before**

```ts
const url = new URL(href.value, import.meta.client ? window.location.href : 'http://localhost')
return {
  path: url.pathname,
  fullPath: url.pathname,
  // ...
}
```

After

```ts
let url: URL
try {
  url = new URL(href.value, import.meta.client ? window.location.href : 'http://localhost')
} catch {
  return undefined
}
return {
  path: url.pathname,
  fullPath: url.pathname,
  // ...
}
```

Solution

Return `undefined` on error. The slot prop route becomes `undefined` and the component handles "no route" gracefully instead of crashing.

---

### 6. `_getPayloadURL()` — payload URL parser (`payload.ts:79`)

**Problem**

Internal function. A crash here produces an unclear error that's hard to debug without tracing the call stack.

**Before**

```ts
const u = new URL(url, 'http://localhost')
```

After

```ts
let u: URL
try {
  u = new URL(url, 'http://localhost')
} catch {
  throw new Error('Invalid payload URL: ' + url)
}
```

Solution

Throw a clear error message that names "payload URL" and shows the value, making root cause immediately identifiable.

---

### Fallback Strategy

| Function | Strategy |
|---|---|
| `navigateTo` | Descriptive Error (user sees what went wrong) |
| `encodeURL` | Return input as-is (transparent, no-op) |
| `<NuxtLink>` prefetch | Return raw path (best-effort) |
| `<NuxtLink>` route getter | Return `undefined` (graceful degrade) |
| `_getPayloadURL` | Descriptive Error (debug-friendly) |

### Test

`packages/nuxt/test/url-try-catch.test.ts` — 15 assertions: `encodeURL` happy path and edge cases, `navigateTo` error behavior, `encodeRoutePath` sanity.

### CODE REVIEW

**Review by danielroe — COMMENTED** [review]
there's a lot of duplicated code here - can you create a util in src/app/utils and reuse it?

**Comment by faisalaffan** [comment]
> there's a lot of duplicated code here - can you create a util in src/app/utils and reuse it?

okay working on it

**Comment by faisalaffan** [comment]
i already add utils. please kindly check @danielroe

**Comment by luc122c** [comment]
I like the idea of not throwing in functions, but I don't think it needs any new utilities creating. Nuxt already depends on [ufo](https://github.com/unjs/ufo) which provides URL parsing utilities.

| PR usage | `ufo` alternative |
|---|---|
| `resolveURL(toPath).protocol` | `parseURL(toPath).protocol` |
| `tryResolveURL(location)` → fallback | `parseURL(location).pathname` |
| `tryResolveURL(path)?.href ?? path` | `stringifyParsedURL(parseURL(path))` |
| `tryResolveURL(href)` → undefined | `parseURL(href).pathname` |
| `resolveURL(url).host` | `parseURL(url).host` |

Same fix, no new utils to maintain. We can just use the URL parser that's already there.

**Comment by faisalaffan** [comment]
Makes sense. Switched to `parseURL`/`stringifyParsedURL` from `ufo` — removed both utilities from `utils.ts`, net -39 lines.

Only thing: for `encodeURL` I kept a fallback `(pathname || search || hash) ? result : location` because `parseURL` returns empty pathname for broken input like `http://a b.com` where `new URL()` would throw.

**Comment by faisalaffan** [comment]
hi @luc122c how about this. i already change to ufo packages ?

File: packages/nuxt/src/app/composables/payload.ts [hunk 1]
```diff
@@ -76,7 +75,7 @@ export function preloadPayload (url: string, opts: LoadPayloadOptions = {}): Pro
 
 const filename = '_payload.json'
 async function _getPayloadURL (url: string, opts: LoadPayloadOptions = {}) {
-  const u = new URL(url, 'http://localhost')
+  const u = parseURL(url, 'http://localhost')
   if (u.host !== 'localhost' || hasProtocol(u.pathname, { acceptRelative: true })) {
     throw new Error('Payload URL must not include hostname: ' + url)
   }
```
Comment (luc122c) [inline packages/nuxt/src/app/composables/payload.ts:81]:
```suggestion
  if (hasProtocol(url, { acceptRelative: true })) {
    throw new Error('Payload URL must not include hostname: ' + url)
  }
```

The imports would need to be updated too.

File: packages/nuxt/src/app/composables/router.ts [hunk 2]
```diff
@@ -156,7 +156,7 @@ export const navigateTo = (to: RouteLocationRaw | undefined | null, options?: Na
 
   // Early open handler
   if (import.meta.client && options?.open) {
-    const { protocol } = new URL(toPath, window.location.href)
+    const { protocol } = parseURL(toPath, window.location.href)
     if (protocol && isScriptProtocol(protocol)) {
       throw new Error(`Cannot navigate to a URL with '${protocol}' protocol.`)
     }
```
Comment (luc122c) [inline packages/nuxt/src/app/composables/router.ts:162]: We don't need to add an additional protocol to check the protocol.
```suggestion
    const { protocol } = parseURL(toPath)
    if (protocol && isScriptProtocol(protocol)) {
      throw new Error(`Cannot navigate to a URL with '${protocol}' protocol.`)
    }
```

File: packages/nuxt/src/app/composables/router.ts [hunk 3]
```diff
@@ -180,7 +180,7 @@ export const navigateTo = (to: RouteLocationRaw | undefined | null, options?: Na
     if (!options?.external) {
       throw new Error('Navigating to an external URL is not allowed by default. Use `navigateTo(url, { external: true })`.')
     }
-    const { protocol } = new URL(toPath, import.meta.client ? window.location.href : 'http://localhost')
+    const { protocol } = parseURL(toPath, import.meta.client ? window.location.href : 'http://localhost')
```
Comment (luc122c) [inline packages/nuxt/src/app/composables/router.ts:183]: We don't need to add a protocol to check for the protocol
```suggestion
    const { protocol } = parseURL(toPath)
```

File: packages/nuxt/src/app/composables/router.ts [hunk 4]
```diff
@@ -341,16 +341,17 @@ export function resolveRouteObject (to: Exclude<RouteLocationRaw, string>): stri
  * @internal
  */
 export function encodeURL (location: string, isExternalHost = false): string {
-  const url = new URL(location, 'http://localhost')
+  const url = parseURL(location, 'http://localhost')
```
Comment (luc122c) [inline packages/nuxt/src/app/composables/router.ts:344]: We don't need the default
```suggestion
  const url = parseURL(location)
```

File: packages/nuxt/src/app/components/nuxt-link.ts [hunk 5]
```diff
@@ -419,7 +419,9 @@ export function defineNuxtLink (options: NuxtLinkOptions): NuxtLinkComponent & R
         const path = typeof to.value === 'string'
           ? to.value
           : isExternal.value ? resolveRouteObject(to.value) : router.resolve(to.value).fullPath
-        const normalizedPath = isExternal.value ? new URL(path, window.location.href).href : path
+        const normalizedPath = isExternal.value
+          ? stringifyParsedURL(parseURL(path, window.location.href))
```
Comment (luc122c) [inline packages/nuxt/src/app/components/nuxt-link.ts:423]: We don't need to provide the base
```suggestion
          ? stringifyParsedURL(parseURL(path))
```

File: packages/nuxt/src/app/components/nuxt-link.ts [hunk 6]
```diff
@@ -527,7 +529,7 @@ export function defineNuxtLink (options: NuxtLinkOptions): NuxtLinkComponent & R
             get route () {
               if (!href.value) { return undefined }
 
-              const url = new URL(href.value, import.meta.client ? window.location.href : 'http://localhost')
+              const url = parseURL(href.value, import.meta.client ? window.location.href : 'http://localhost')
```
Comment (luc122c) [inline packages/nuxt/src/app/components/nuxt-link.ts:532]: We don't need to provide the base.
```suggestion
              const url = parseURL(href.value)
```

**Review by luc122c — COMMENTED** [review]
[parseURL](https://github.com/unjs/ufo/tree/v1.6.4#parseurlinput-defaultproto) doesn't have the same signature as `URL`; we don't need still provide a base. I've suggested some changes, but I haven't checked that the tests still pass.

Also, `url-try-catch.test.ts` could probably be renamed to `url-utils.test.ts` or something like that.
